README.md
<p align="center">
  <a href="https://github.com/Fliggy-Mobile">
    <img width="200" src="https://gw.alicdn.com/tfs/TB1a288sxD1gK0jSZFKXXcJrVXa-360-360.png">
  </a>
</p>

<h1 align="center">ftoast</h1>


<div align="center">

<p>Simple, beautiful Toast</p>

<p>Help developers create flexible, concise and beautiful Toast</p>

<p><strong>Author：<a href="https://github.com/chenBingX">Newton</a>(<a href="coorchice.cb@alibaba-inc.com">coorchice.cb@alibaba-inc.com</a>)</strong></p>
<p>

<a href="https://pub.dev/packages/ftoast#-readme-tab-">
    <img height="20" src="https://img.shields.io/badge/Version-2.0.0-important.svg">
</a>


<a href="https://github.com/Fliggy-Mobile/ftoast">
    <img height="20" src="https://img.shields.io/badge/Build-passing-brightgreen.svg">
</a>


<a href="https://github.com/Fliggy-Mobile">
    <img height="20" src="https://img.shields.io/badge/Team-FAT-ffc900.svg">
</a>

<a href="https://www.dartcn.com/">
    <img height="20" src="https://img.shields.io/badge/Language-Dart-blue.svg">
</a>

<a href="https://pub.dev/documentation/ftoast/latest/ftoast/ftoast-library.html">
    <img height="20" src="https://img.shields.io/badge/API-done-yellowgreen.svg">
</a>

<a href="https://www.apache.org/licenses/LICENSE-2.0.txt">
   <img height="20" src="https://img.shields.io/badge/License-Apache--2.0-blueviolet.svg">
</a>

<p>
<p>

</div>

![](https://gw.alicdn.com/tfs/TB1CdJ5ffzO3e4jSZFxXXaP_FXa-1466-562.png)


**English | [简体中文](https://github.com/Fliggy-Mobile/ftoast/blob/master/README_CN.md)**

> Like it? Please cast your **Star**  🥰 ！

# ✨ Features

- Provide a convenient way to create **Toast**

- Support primary information, secondary information, and customize their style

- Support flexible icon view

- Support to configure corner size and background color

- Support queue display

- Support custom **Toast** style


# 🛠 Guide

In FToast, developers can easily create a **Toast** with simple code.

```dart
FToast.toast(context, msg: "FToast");
```

## ⚙️ Parameters

|Param|Type|Necessary|Default|desc|
|---|---|:---:|---|---|
|context|BuildContext|true|null|page environment|
|toast|Widget|false|null|custom toast view, will override the default view|
|msg|String|false|null|main title|
|msgStyle|TextStyle|false|null|main title text style|
|subMsg|String|false|null|Subtitle|
|subMsgStyle|TextStyle|false|null|Subtitle text style|
|subMsgSpace|double|false|12.0|Subtitle and main title spacing|
|corner|double|false|6.0|Corner|
|color|Color|false|Colors.black54|Color|
|duration|double|false|1800|Impression duration|
|padding|EdgeInsets|false|`EdgeInsets.only(left: 16.0, right: 16.0, top: 16.0, bottom: 16.0)`|inner spacing|
|image|Widget|false|null|icon|
|imageDirection|AxisDirection|false|AxisDirection.up|position of icon relative to text|
|imageSpace|double|false|9.0|distance between icon and text|

## 🔩 Basic

![](https://gw.alicdn.com/tfs/TB1YG.Mc8FR4u4jSZFPXXanzFXa-365-720.gif)

```dart
FToast.toast(
  context,

  /// 配置显示时长
  ///
  /// Configure display duration
  duration: 800,

  /// 配置 Msg
  ///
  /// set Msg
  msg: "I'm FToast: ${count++}",

  /// 配置 Msg 样式
  ///
  /// set Msg style
  msgStyle: TextStyle(color: Colors.white),
)
```

## 🔆 SubMsg

![](https://gw.alicdn.com/tfs/TB1b0.5NQT2gK0jSZFkXXcIQFXa-365-720.gif)

```dart
FToast.toast(
  context,
  msg: "This is Msg",

  /// 配置 subMsg
  ///
  /// set subMsg
  subMsg: "Welcome to use FToast. This is subMsg!",

  /// 配置 SubMsg 样式
  ///
  /// set SubMsg style
  subMsgStyle: TextStyle(color: Colors.white, fontSize: 13),
)
```

## 🔳 Image

![](https://gw.alicdn.com/tfs/TB13JPNbZVl614jSZKPXXaGjpXa-365-720.gif)

```dart
FToast.toast(
  context,
  msg: "This is Msg",
  subMsg: "Welcome to use FToast. This is subMsg!",

  /// 配置图标
  ///
  /// set image
  image: Icon(
    Icons.star,
    color: Colors.yellow,
  ),

  /// 配置图标相对文本的位置
  ///
  /// set position of icon relative to text
  imageDirection: up,
)
```


##  🔩 Custom style

![](https://gw.alicdn.com/tfs/TB1wlI_NUT1gK0jSZFrXXcNCXXa-365-720.gif)

```dart
FToast.toast(
  context,
  /// 自定义 Toast 样式
  /// 
  /// Custom Toast style
  toast: FSuper(
    text: "Custom Toast",
    style: TextStyle(color: Colors.grey),
    padding: EdgeInsets.all(12),
    shadowColor: Colors.yellow,
    shadowBlur: 80,
  ),
)
```


# 😃 How to use？

Add dependencies in the project `pubspec.yaml` file:

## 🌐 pub dependency

```
dependencies:
  ftoast: ^<version number>
```

> ⚠️ Attention，please go to [**pub**] (https://pub.dev/packages/ftoast) to get the latest version number of **FToast**

## 🖥 Git dependency

```
dependencies:
  ftoast:
    git:
      url: 'git@github.com:Fliggy-Mobile/ftoast.git'
      ref: '<Branch number or tag number>'
```

> ⚠️ Attention，please refer to [**FToast**] (https://github.com/Fliggy-Mobile/ftoast) official project for branch number or tag.


# 💡 License

```
Copyright 2020-present Fliggy Android Team <alitrip_android@list.alibaba-inc.com>.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at following link.

    https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

```


### Like it? Please cast your [**Star**](https://github.com/Fliggy-Mobile/ftoast) 🥰 ！


---

# How to run Demo project?

1. **clone** project to local

2. Enter the project `example` directory and run the following command

```
flutter create .
```

3. Run the demo in `example`





CHANGELOG.md
## 2.0.0
- Support null-safety

## 1.0.0

- the first version


lib/ftoast.dart
import 'dart:async';
import 'dart:collection';
import 'dart:ui';

import 'package:flutter/material.dart';

/// 帮助开发者创建灵活的，简洁的，精美的 Toast
///
/// Help developers create flexible, concise and beautiful Toast
///
/// e.g.:
/// ```
/// FToast.toast(context, msg: "Hi, FWidget");
class FToast {
  static Queue<_ToastData>? _entryQueue;
  static _ToastData? _current;

  /// [context] - 页面环境
  /// [toast] - 自定义 toast 视图，会覆盖默认视图
  /// [msg] - 主信息
  /// [msgStyle] - 主信息文本样式
  /// [subMsg] - 副信息
  /// [subMsgStyle] - 副信息文本样式
  /// [subMsgSpace] - 副信息与主信息的间距
  /// [corner] - 边角
  /// [color] - 颜色
  /// [duration] - 展示时长
  /// [padding] - 内间距
  /// [image] - 图标
  /// [imageDirection] - 图标相对文本的位置
  /// [imageSpace] - 图标与文本的间距
  ///
  /// [context]-page environment
  /// [toast]-custom toast view, will override the default view
  /// [msg]-main title
  /// [msgStyle]-main title text style
  /// [subMsg ]-Subtitle
  /// [subMsgStyle]-Subtitle text style
  /// [subMsgSpace]-Subtitle and main title spacing
  /// [corner]-Corner
  /// [color]-Color
  /// [duration]- Impression duration
  /// [padding]-inner spacing
  /// [image]-icon
  /// [imageDirection]-position of icon relative to text
  /// [imageSpace]-distance between icon and text
  static toast(
    final BuildContext context, {
    final Widget? toast,
    final String msg = "",
    final TextStyle? msgStyle,
    final String? subMsg,
    final TextStyle? subMsgStyle,
    final double subMsgSpace = 12.0,
    final double corner = 6.0,
    final Color color = Colors.black54,
    final int duration = 1800,
    final EdgeInsets padding =
        const EdgeInsets.only(left: 16.0, right: 16.0, top: 16.0, bottom: 16.0),
    final Widget? image,
    final AxisDirection imageDirection = AxisDirection.up,
    final double imageSpace = 9.0,
  }) {
    Widget buildToast() {
      TextStyle defaultStyle = TextStyle(
        color: Colors.white,
        fontSize: 16,
        height: 1.0,
      );
      List<Widget> children = [];

      if (image != null && imageDirection == AxisDirection.up) {
        children.add(image);
        children.add(SizedBox(height: imageSpace));
      }

      /// msg
      if (image != null && imageDirection == AxisDirection.left) {
        children.add(Row(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            image,
            SizedBox(width: imageSpace),
            Text(
              msg,
              style: msgStyle ?? defaultStyle,
            )
          ],
        ));
      } else if (image != null && imageDirection == AxisDirection.right) {
        children.add(Row(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Text(
              msg,
              style: msgStyle ?? defaultStyle,
            ),
            SizedBox(width: imageSpace),
            image,
          ],
        ));
      } else {
        children.add(Text(
          msg,
          style: msgStyle ?? defaultStyle,
        ));
      }

      /// subMsg
      if (subMsg != null) {
        children.add(
          SizedBox(height: subMsgSpace),
        );
        children.add(Text(
          subMsg,
          style: subMsgStyle ?? defaultStyle.copyWith(fontSize: 13),
        ));
      }
      if (image != null && imageDirection == AxisDirection.down) {
        children.add(SizedBox(height: imageSpace));
        children.add(image);
      }
      return ConstrainedBox(
        constraints: BoxConstraints(
            minWidth: 80.0,
            minHeight: 38.0,
            maxWidth: MediaQueryData.fromWindow(window).size.width - 48.0),
        child: Container(
          padding: padding,
          decoration: BoxDecoration(
            color: color,
            borderRadius: BorderRadius.all(Radius.circular(corner)),
          ),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: children,
          ),
        ),
      );
    }

    if (_entryQueue == null) _entryQueue = Queue();
    OverlayEntry entry = OverlayEntry(builder: (context) {
      return IgnorePointer(
        child: Material(
          color: Colors.transparent,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              toast ?? buildToast(),
            ],
          ),
        ),
      );
    });
    _ToastData toastData = _ToastData()
      ..context = context
      ..entry = entry
      ..duration = duration;
    _entryQueue?.addLast(toastData);
    _show();
  }

  static _show() {
    if (_current == null) {
      final _ToastData? first = (_entryQueue?.isNotEmpty ?? false)
          ? _entryQueue?.removeFirst()
          : null;
      _current = first;
      if (first != null) {
        try {
          if (first.context != null &&
              (first.context!.findRenderObject()?.attached ?? false) &&
              first.entry != null) {
            OverlayState? overlayState = Overlay.of(first.context!);
            overlayState?.insert(first.entry!);
            first.showed = true;
            Timer(Duration(milliseconds: first.duration), () {
              first.dispose();
              _current = null;
              _show();
            });
          } else {
            first.dispose();
            _current = null;
          }
        } catch (e) {
          _current = null;
        }
      }
    }
  }
}

class _ToastData {
  OverlayEntry? entry;
  int duration = 1800;
  BuildContext? context;
  bool showed = false;

  dispose() {
    if (showed) entry?.remove();
    entry = null;
    context = null;
  }
}


example/lib/main.dart
import 'dart:ui';

import 'package:fcommon/fcommon.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:fsuper/fsuper.dart';
import 'package:ftoast/ftoast.dart';
import 'package:fcontrol/fdefine.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  int count = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Builder(builder: (context) {
        var appBar = AppBar(
          backgroundColor: mainBackgroundColor,
          title: const Text(
            'FToast',
            style: TextStyle(color: mainTextTitleColor),
          ),
          centerTitle: true,
        );
        var shadowTop = MediaQueryData.fromWindow(window).size.height -
            appBar.preferredSize.height -
            20;
        return Scaffold(
          backgroundColor: mainBackgroundColor,
          appBar: appBar,
          body: Stack(
            children: [
              SingleChildScrollView(
                child: Container(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: [
                      buildTitle("FToast"),
                      buildMiddleMargin(),
                      buildMiddleMargin(),
                      buildMiddleMargin(),

                      /// FToast
                      buildFToast(context),
                      const SizedBox(height: 500),
                      buildTitle("SubMsg"),
                      buildMiddleMargin(),
                      buildMiddleMargin(),
                      buildMiddleMargin(),

                      /// SubMsg
                      buildSubMsg(context),
                      const SizedBox(height: 500),

                      buildTitle("Image"),
                      buildMiddleMargin(),
                      buildMiddleMargin(),
                      buildMiddleMargin(),

                      /// Image
                      buildImage(context),
                      const SizedBox(height: 500),

                      buildTitle("Custom Toast"),
                      buildMiddleMargin(),
                      buildMiddleMargin(),
                      buildMiddleMargin(),

                      /// Custom Toast
                      buildCustomToast(context),
                      const SizedBox(height: 500),

                      buildBiggestMargin(),
                      buildBiggestMargin(),
                    ],
                  ),
                ),
              ),
              FSuper(
                width: double.infinity,
                height: 26.0,
                margin: EdgeInsets.only(top: shadowTop),
                gradient: LinearGradient(
                    begin: Alignment.topCenter,
                    end: Alignment.bottomCenter,
                    colors: [
                      Colors.transparent,
                      Colors.black26,
                    ]),
              ),
            ],
          ),
        );
      }),
    );
  }

  Widget buildFToast(BuildContext context) {
    return FSuper(
      text: "Click to show FToast",
      style: TextStyle(color: mainTextTitleColor),
      backgroundColor: mainBackgroundColor,
      padding: EdgeInsets.all(12.0),
      onClick: () {
        FToast.toast(
          context,

          /// 配置显示时长
          ///
          /// Configure display duration
          duration: 800,

          /// 配置 Msg
          ///
          /// set Msg
          msg: "I'm FToast: ${count++}",

          /// 配置 Msg 样式
          ///
          /// set Msg style
          msgStyle: TextStyle(color: Colors.white),
        );
      },
      corner: FCorner.all(20),
      isSupportNeumorphism: true,
    );
  }

  Widget buildSubMsg(BuildContext context) {
    return FSuper(
      text: "Click to show FToast",
      style: TextStyle(color: mainTextTitleColor),
      backgroundColor: mainBackgroundColor,
      padding: EdgeInsets.all(12.0),
      onClick: () {
        FToast.toast(
          context,
          msg: "This is Msg",

          /// 配置 subMsg
          ///
          /// set subMsg
          subMsg: "Welcome to use FToast. This is subMsg!",

          /// 配置 SubMsg 样式
          ///
          /// set SubMsg style
          subMsgStyle: TextStyle(color: Colors.white, fontSize: 13),
        );
      },
      corner: FCorner.all(20),
      isSupportNeumorphism: true,
    );
  }

  AxisDirection up = AxisDirection.up;

  Widget buildImage(BuildContext context) {
    return FSuper(
      text: "Click to show FToast",
      style: TextStyle(color: mainTextTitleColor),
      backgroundColor: mainBackgroundColor,
      padding: EdgeInsets.all(12.0),
      onClick: () {
        switch (up) {
          case AxisDirection.up:
            up = AxisDirection.left;
            break;
          case AxisDirection.left:
            up = AxisDirection.right;
            break;
          case AxisDirection.right:
            up = AxisDirection.down;
            break;
          case AxisDirection.down:
            up = AxisDirection.up;
            break;
        }
        FToast.toast(
          context,
          msg: "This is Msg",
          subMsg: "Welcome to use FToast. This is subMsg!",

          /// 配置图标
          ///
          /// set image
          image: Icon(
            Icons.star,
            color: Colors.yellow,
          ),

          /// 配置图标相对文本的位置
          ///
          /// set position of icon relative to text
          imageDirection: up,
        );
      },
      corner: FCorner.all(20),
      isSupportNeumorphism: true,
    );
  }

  Widget buildCustomToast(BuildContext context) {
    return FSuper(
      text: "Click to show FToast",
      style: TextStyle(color: mainTextTitleColor),
      backgroundColor: mainBackgroundColor,
      padding: EdgeInsets.all(12.0),
      onClick: () {
        FToast.toast(
          context,

          /// 自定义 Toast 样式
          ///
          /// Custom Toast style
          toast: FSuper(
            text: "Custom Toast",
            style: TextStyle(color: Colors.grey),
            padding: EdgeInsets.all(12),
            shadowColor: Colors.yellow,
            shadowBlur: 80,
          ),
        );
      },
      corner: FCorner.all(20),
      isSupportNeumorphism: true,
    );
  }
}

class Page_2 extends StatefulWidget {
  @override
  _Page_2State createState() => _Page_2State();
}

class _Page_2State extends State<Page_2> {
  @override
  Widget build(BuildContext context) {
    return Container(
      alignment: Alignment.center,
      color: mainBackgroundColor,
      child: FSuper(
        width: 100,
        height: 80,
        backgroundColor: mainBackgroundColor,
        onClick: () {
          FToast.toast(context, msg: "Hi");
          Navigator.pop(context);
        },
        isSupportNeumorphism: true,
      ),
    );
  }
}
